cpp/src/sort/segmented_sort.cu

```diff
@@ -34,6 +39,195 @@ namespace {
+                                        begin_offsets,
+                                        end_offsets,
+                                        stream.value());
+  }
+
+  template <typename KeyT, typename ValueT, typename OffsetIteratorT>
+  void fast_sort_descending(KeyT const* keys_in,
+                            KeyT* keys_out,
+                            ValueT const* values_in,
+                            ValueT* values_out,
+                            int num_items,
+                            int num_segments,
+                            OffsetIteratorT begin_offsets,
+                            OffsetIteratorT end_offsets,
+                            rmm::cuda_stream_view stream)
+  {
+    rmm::device_buffer d_temp_storage;
+    size_t temp_storage_bytes = 0;
+    // DeviceSegmentedSort is faster then DeviceSegmentedRadixSort at this time
+    cub::DeviceSegmentedSort::SortPairsDescending(d_temp_storage.data(),
+                                                  temp_storage_bytes,
+                                                  keys_in,
+                                                  keys_out,
+                                                  values_in,
+                                                  values_out,
+                                                  num_items,
+                                                  num_segments,
+                                                  begin_offsets,
+                                                  end_offsets,
+                                                  stream.value());
+    d_temp_storage = rmm::device_buffer{temp_storage_bytes, stream};
+
+    cub::DeviceSegmentedSort::SortPairsDescending(d_temp_storage.data(),
+                                                  temp_storage_bytes,
+                                                  keys_in,
+                                                  keys_out,
+                                                  values_in,
+                                                  values_out,
+                                                  num_items,
+                                                  num_segments,
+                                                  begin_offsets,
+                                                  end_offsets,
+                                                  stream.value());
+  }
+
+  template <typename T>
+  void fast_sort(column_view const& input,
+                 column_view const& segment_offsets,
+                 mutable_column_view& indices,
+                 bool ascending,
+                 rmm::cuda_stream_view stream)
+  {
+    // CUB's segmented sort functions cannot accept iterators.
+    // We create a temporary column here for it to use.
+    auto temp_col =
+      cudf::detail::allocate_like(input, input.size(), mask_allocation_policy::NEVER, stream);
+    mutable_column_view output_view = temp_col->mutable_view();
+
+    if (ascending) {
```

Comment:
We should reduce a lot of this code by using a lambda with perfect forwarding to select the CUB function. All the input parameters and temp storage appear to be the same. Example: https://godbolt.org/z/3sY39EMYz

With that, you can eliminate `fast_sort_ascending/fast_sort_descending` and keep the implementation unified in `fast_sort`.

Reply:
I find this a bit frustrating. There is technically not a lot of code here. It only appears that way because of the clang-formatting. Your suggestion actually only reduces the number of lines of code from 12 to 10. The `std::forward` helps to reduce the source text lines count (dramatically) because the parameter list does not need to be expanded by clang-format. So I feel this is more of a solution to get around code formatting than to make the code more clear.

Reply:
Maybe I'm missing something, but I think my proposal would reduce the common logic by a substantial amount, irrespective of clang-formatting? I can push a commit to my fork if you want to see exactly what I'm thinking (but probably tomorrow).

Reply:
I pushed a change. Let me know what you think. Maybe I misunderstood.

Reply:
This change is great! It reduced the number of times that arguments were specified to 1 time instead of 4 times (for each ascending/descending, for both calls to the function). Thank you.

Reply:
I'm getting this error in CI but not locally
```
21:14:43 $SRC_DIR/cpp/src/sort/segmented_sort.cu(85): error #177-D: parameter "args" was declared but never referenced
```
I could use some help figuring out why this is happening and how to fix it.
https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-cpu-cuda-build/CUDA=11.5/13331/console

Reply:
Adding `[[maybe_unused]]` to the declaration seems to have fixed it.